cxl/lib/libcxl.c

```diff
@@ -14541,3 +14541,148 @@ CXL_EXPORT int cxl_memdev_ddr_page_select_get(struct cxl_memdev *memdev)
+    int rc = 0;
+    if (refresh_select_option != 0 && refresh_select_option != 2 && refresh_select_option != 4) {
+        fprintf(stderr, "Error: Invalid refresh select option value: %d\n", refresh_select_option);
+        return -EINVAL;
+    }
+
+    cmd = cxl_cmd_new_raw(memdev, CXL_MEM_COMMAND_ID_DDR_REFRESH_MODE_SET_OPCODE);
+    if (!cmd) {
+        fprintf(stderr, "%s: cxl_cmd_new_raw returned Null output\n",
+                cxl_memdev_get_devname(memdev));
+        return -ENOMEM;
+    }
+
+    query = cmd->query_cmd;
+    cinfo = &query->commands[cmd->query_idx];
+
+    /* update payload size */
+    cinfo->size_in = CXL_MEM_COMMAND_ID_DDR_REFRESH_MODE_SELECT_SET_PAYLOAD_IN_SIZE;
+    if (cinfo->size_in > 0) {
+        cmd->input_payload = calloc(1, cinfo->size_in);
+        if (!cmd->input_payload)
+            return -ENOMEM;
+        cmd->send_cmd->in.payload = (u64)cmd->input_payload;
+        cmd->send_cmd->in.size = cinfo->size_in;
+    }
+
+    handle_refresh_selection_in = (void *) cmd->send_cmd->in.payload;
+
+    handle_refresh_selection_in->refresh_select.ddr_refresh_val = refresh_select_option;
+
+    rc = cxl_cmd_submit(cmd);
+    if (rc < 0) {
+        fprintf(stderr, "%s: cmd submission failed: %d (%s)\n",
+                cxl_memdev_get_devname(memdev), rc, strerror(-rc));
+        goto out;
+    }
+
+    rc = cxl_cmd_get_mbox_status(cmd);
+    if (rc != 0) {
+        fprintf(stderr, "%s: firmware status: %d\n",
+                cxl_memdev_get_devname(memdev), rc);
+        rc = -ENXIO;
+        goto out;
+    }
+
+    if (cmd->send_cmd->id != CXL_MEM_COMMAND_ID_DDR_REFRESH_MODE_SET) {
+        fprintf(stderr, "%s: invalid command id 0x%x (expecting 0x%x)\n",
+                cxl_memdev_get_devname(memdev), cmd->send_cmd->id,
+                CXL_MEM_COMMAND_ID_DDR_REFRESH_MODE_SET);
+        return -EINVAL;
+    }
+
+out:
+    cxl_cmd_unref(cmd);
+    return rc;
+}
+
+/* DDR REFRESH MODE GET */
+#define CXL_MEM_COMMAND_ID_CXL_DDR_REFRESH_MODE_GET CXL_MEM_COMMAND_ID_RAW
+#define CXL_MEM_COMMAND_ID_CXL_DDR_REFRESH_MODE_GET_OPCODE 0xFB33
+
+struct cxl_mbox_handle_ddr_refresh_select_out {
+  struct ddr_refresh_mode refresh_select;
+} __attribute__((packed));
+
+CXL_EXPORT int cxl_memdev_ddr_refresh_mode_get(struct cxl_memdev *memdev)
+{
+    struct cxl_cmd *cmd;
+    struct cxl_mem_query_commands *query;
+    struct cxl_command_info *cinfo;
+    struct  cxl_mbox_handle_ddr_refresh_select_out *handle_refresh_selection_out;
+    int rc = 0;
+
+    cmd = cxl_cmd_new_raw(memdev,  CXL_MEM_COMMAND_ID_CXL_DDR_REFRESH_MODE_GET_OPCODE);
+    if (!cmd) {
+        fprintf(stderr, "%s: cxl_cmd_new_raw returned Null output\n",
+                cxl_memdev_get_devname(memdev));
+        return -ENOMEM;
+    }
+
+    query = cmd->query_cmd;
+    cinfo = &query->commands[cmd->query_idx];
+
+    /* used to force correct payload size */
+    cinfo->size_in = CXL_MEM_COMMAND_ID_LOG_INFO_PAYLOAD_IN_SIZE;
```

Comment:
I think it is okay to set to 0 here as we dont have any payload in a `get` command.

Reply:
Updated